components/protocol-wasm/src/lib.rs

```diff
@@ -0,0 +1,213 @@
+use chrono::{DateTime, Utc};
+use base64::{Engine as _, engine::{self, general_purpose}};
+use std::str::FromStr; // 0.4.15
+
+use bindings::protocol as protocolbindings;
+struct Component;
+
+pub use semver::{Version, VersionReq};
+
+use anyhow::Error;
+use anyhow::anyhow;
+
+use warg_protocol::{
+  package,
+  proto_envelope::{ProtoEnvelope, ProtoEnvelopeBody}, 
+  SerdeEnvelope,
+  registry::{MapCheckpoint, RecordId, LogId, LogLeaf, MapLeaf},
+};
+use warg_crypto::{signing, Decode, hash::{Hash, Sha256, HashAlgorithm, DynHash}};
+use warg_transparency::{log::LogProofBundle, map::MapProofBundle};
+use warg_api::proof::ProofError;
+
+bindings::export!(Component);
+/// Represents information about a registry package.
+#[derive(Debug, Clone)]
+// #[serde(rename_all = "camelCase")]
+pub struct PackageInfo {
+    /// The name of the package.
+    pub name: String,
+    /// The last known checkpoint of the package.
+    // #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub checkpoint: Option<String>,
+    // pub checkpoint: Option<String>,
+    /// The current validation state of the package.
+    // #[serde(default)]
+    pub state: package::Validator,
+}
+
+impl PackageInfo {
+    /// Creates a new package info for the given package name and url.
+    pub fn new(name: impl Into<String>, 
+    ) -> Self {
+        Self {
+            name: name.into(),
+            checkpoint: None,
+            state: package::Validator::default(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ProtoBody(protocolbindings::ProtoEnvelopeBody);
+
+impl<Content> TryFrom<ProtoBody> for ProtoEnvelope<Content>
+where
+    Content: Decode,
+{
+    type Error = Error;
+
+    fn try_from(value: ProtoBody) -> Result<Self, Self::Error> {
+        let contents = Content::decode(&value.0.content_bytes)?;
+        let envelope = ProtoEnvelope {
+            contents,
+            content_bytes: value.0.content_bytes,
+            key_id: value.0.key_id.into(),
+            signature: signing::Signature::from_str(&value.0.signature).unwrap(),
+        };
+        Ok(envelope)
+    }
+}
+
+fn perm_binding(permission: &package::model::Permission) -> protocolbindings::Permission {
+    match permission {
+        &package::Permission::Release => protocolbindings::Permission::Release,
+        &package::Permission::Yank => protocolbindings::Permission::Yank,
+        &_ => protocolbindings::Permission::Release,
+    }
+}
+
+impl protocolbindings::Protocol for Component {
+    fn prove_inclusion(input: protocolbindings::Inclusion, checkpoint: protocolbindings::MapCheckpoint, heads: Vec<protocolbindings::LogLeaf>) {
+      let map_checkpoint = MapCheckpoint {
```

Comment:
Dead code, value `map_checkpoint` is not used.

Reply:
deleted